File: Matlab_heatEqn_1d/mgrit.m
function [u,v,g] = mgrit(l, L, u, v, g, Phi, app, m, t, iter)

if l == L
    %% coarse-grid solve (or L=1): time stepping
    for i=2:length(t{l})
        u{l}(:,i) = g{l}(:,i) + ...
                     feval(Phi,u{l}(:,i-1),t{l}(i),t{l}(i-1),app,l);
    end
else
    % MGRIT algorithm
    %% F- and C-points on current grid level
    nt           = length(t{l});
    Cpts         = 1:m:nt;
    Fpts         = 1:nt;
    Fpts(Cpts)   = [];

    % fprintf(1,'********* start MGRIT *********\n');
    % fprintf(1,'  l                  = %d\n', l);
    % fprintf(1,'  nt                 = %d\n', nt);
    % fprintf(1,'  m                  = %d\n', m);
    % fprintf(1,'  number of F-points = %d\n', length(Fpts));
    % fprintf(1,'  number of C-points = %d\n', length(Cpts));


    %% F-relaxation
    if ((l > 1) || ( (iter == 1) && (l == 1) ))
        for i=Fpts
            u{l}(:,i) = g{l}(:,i) + ...
                         feval(Phi,u{l}(:,i-1),t{l}(i),t{l}(i-1),app,l);
        end
    end

    %% C-relaxation
    for i=Cpts(2:end)
       u{l}(:,i) = g{l}(:,i) + ...
                    feval(Phi,u{l}(:,i-1),t{l}(i),t{l}(i-1),app,l);
    end

    %% F-relaxation
    for i=Fpts
       u{l}(:,i) = g{l}(:,i) + ...
                    feval(Phi,u{l}(:,i-1),t{l}(i),t{l}(i-1),app,l);
    end

    %% inject the fine-grid approximation and compute FAS right-hand side
    v{l+1} = u{l}(:,Cpts);
    % FAS right-hand side: R(g) + A_c( R(u) ) - R( A(u) )
    for i=2:length(Cpts)
        g{l+1}(:,i) = g{l}(:,Cpts(i)) + ...
                       v{l+1}(:,i) - ...
                       feval(Phi,v{l+1}(:,i-1),t{l+1}(i),t{l+1}(i-1), ...
                             app,l+1) - ...
                       u{l}(:,Cpts(i)) + ...
                       feval(Phi,u{l}(:,Cpts(i)-1),t{l}(Cpts(i)), ...
                             t{l}(Cpts(i)-1),app,l);
    end

    %% next level
    u{l+1}  = v{l+1};
    [u,v,g] = mgrit(l+1, L, u, v, g, Phi, app, m, t, iter);
    
    %% correct the approximation u at C-points
    u{l}(:,Cpts) = u{l}(:,Cpts) + (u{l+1}-v{l+1});
    
    %% carry out F-relax to correct the approximation u at F-points
    for i=Fpts
        u{l}(:,i) = g{l}(:,i) + ...
                     feval(Phi,u{l}(:,i-1),t{l}(i),t{l}(i-1),app,l);
    end
end

File: PythonTemplates/mgrit.py
import numpy as np


def mgrit(l, L, u, v, g, m, t, Phi, app):
    # Solve
    if l == L - 1:
        for i in range(1, np.size(t[l])):
            u[l][i] = g[l][i] + Phi(u[l][i - 1], t[l][i - 1], t[l][i], app[l])
        return u, v, g

    # Compute F and C points
    nt = np.size(t[l])
    cpts = np.array(range(0, nt, m))
    fpts = np.array(list(set(np.array(range(0, nt))) - set(cpts)))

    # FCF-relaxation
    for i in np.nditer(fpts):
        u[l][i] = g[l][i] + Phi(u[l][i - 1], t[l][i - 1], t[l][i], app[l])

    for i in np.nditer(cpts[1:]):
        u[l][i] = g[l][i] + Phi(u[l][i - 1], t[l][i - 1], t[l][i], app[l])

    for i in np.nditer(fpts):
        u[l][i] = g[l][i] + Phi(u[l][i - 1], t[l][i - 1], t[l][i], app[l])

    v[l + 1] = np.copy(u[l][cpts])

    for i in range(1, np.size(cpts)):
        g[l + 1][i] = g[l][cpts[i]] \
                      + Phi(u[l][cpts[i] - 1], t[l][cpts[i] - 1], t[l][cpts[i]], app[l]) - u[l][cpts[i]] \
                      - Phi(v[l + 1][i - 1], t[l + 1][i - 1], t[l + 1][i], app[l + 1]) + v[l + 1][i]

    u[l + 1] = np.copy(v[l + 1])

    u, v, g = mgrit(l + 1, L, u, v, g, m, t, Phi, app)

    e = u[l + 1] - v[l + 1]

    u[l][cpts] = u[l][cpts] + e

    # F-relaxation
    for i in np.nditer(fpts):
        u[l][i] = g[l][i] + Phi(u[l][i - 1], t[l][i - 1], t[l][i], app[l])

    return u, v, g
